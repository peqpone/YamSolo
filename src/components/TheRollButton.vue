<script setup lang="ts">
import useDiceStore from '@/stores/dice';
import useGameStore from '@/stores/game';
import { computed } from 'vue';

const diceStore = useDiceStore();
const gameStore = useGameStore();

const canRoll = computed(() => gameStore.canRoll);
const die = () => Math.floor(Math.random() * (6 - 1 + 1) + 1);
const roll = () => {
  if (!canRoll.value) {
    console.debug('Cannot roll the dice');
    return;
  }
  const newDice = [die(), die(), die(), die(), die()];
  diceStore.saveDice(newDice);
  gameStore.increaseAttempts();
};
</script>

<template>
  <div class="button-area">
    <a
      class="button"
      :class="{ disabled: !canRoll }"
      @click="roll"
      @keydown="roll"
    >
      Roll
    </a>
  </div>
</template>

<style lang="scss" scoped>
.button-area {
  display: flex;
  align-items: stretch;
  text-align: center;
  justify-content: center;
  .button {
    display: block;
    width: 100%;
    line-height: 5rem;
    font-size: 3rem;
    font-weight: 100;
    &.disabled {
      cursor: not-allowed;
      opacity: 0.4;
    }
  }
}
</style>
