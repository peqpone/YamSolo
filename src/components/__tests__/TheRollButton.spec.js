import {
  describe, it, expect, beforeAll, vi, afterEach,
} from 'vitest';

import { shallowMount } from '@vue/test-utils';
import { createPinia, setActivePinia } from 'pinia';
import TheRollButton from '../TheRollButton.vue';
import useDiceStore from '../../stores/dice';
import useGameStore from '../../stores/game';

describe('TheRollButton', () => {
  let diceStore;
  let gameStore;
  beforeAll(() => {
    setActivePinia(createPinia());
    diceStore = useDiceStore();
    gameStore = useGameStore();
    diceStore.reset();
    gameStore.reset();
    vi.spyOn(diceStore, 'saveDice');
    vi.spyOn(gameStore, 'increaseAttempts');
  });
  afterEach(() => {
    vi.clearAllMocks();
  });
  describe('Integration test', () => {
    let wrapper;
    it('should correctly mount initial state', () => {
      wrapper = shallowMount(TheRollButton);
      expect(wrapper.html()).toMatchSnapshot();
    });
    it('should disable the button', async () => {
      await gameStore.$patch({ game: { currentAttempt: 3 } });
      expect(wrapper.html()).toMatchSnapshot();
    });
    it('click on the button should not roll the dices', async () => {
      const button = wrapper.find('.main-cta a');
      await button.trigger('click');
      expect(diceStore.saveDice).not.toHaveBeenCalled();
      expect(gameStore.increaseAttempts).not.toHaveBeenCalled();
    });
    it('click on the button should save dice and increase attempts counter', async () => {
      await gameStore.$patch({ game: { currentAttempt: 1 } });
      const button = wrapper.find('.main-cta a');
      await button.trigger('click');
      expect(diceStore.saveDice).toHaveBeenCalledWith([
        { isLocked: false, value: expect.any(Number) },
        { isLocked: false, value: expect.any(Number) },
        { isLocked: false, value: expect.any(Number) },
        { isLocked: false, value: expect.any(Number) },
        { isLocked: false, value: expect.any(Number) },
      ]);
      expect(gameStore.increaseAttempts).toHaveBeenCalled();
    });
    it('click on the button should save dice and increase attempts counter', async () => {
      await gameStore.$patch({ game: { currentAttempt: 1 } });
      await diceStore.$patch({
        rawDice: [
          { value: 1, isLocked: false },
          { value: 1, isLocked: true },
          { value: 1, isLocked: false },
          { value: 1, isLocked: true },
          { value: 1, isLocked: false },
        ],
      });
      const button = wrapper.find('.main-cta a');
      await button.trigger('click');
      expect(diceStore.saveDice).toHaveBeenCalledWith([
        { isLocked: false, value: expect.any(Number) },
        { isLocked: true, value: 1 },
        { isLocked: false, value: expect.any(Number) },
        { isLocked: true, value: 1 },
        { isLocked: false, value: expect.any(Number) },
      ]);
      expect(gameStore.increaseAttempts).toHaveBeenCalled();
    });
  });
});
