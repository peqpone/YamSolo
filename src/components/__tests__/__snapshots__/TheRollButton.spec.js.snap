// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TheRollButton > Integration test > should correctly mount initial state 1`] = `"<div class=\\"main-cta\\"><a class=\\"button\\"> Roll </a></div>"`;

exports[`TheRollButton > Integration test > should disable the button 1`] = `"<div class=\\"main-cta\\"><a class=\\"button disabled\\"> Roll </a></div>"`;
